sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java

```diff
@@ -64,53 +69,92 @@ public GeneratedColumns(InputFactory inputFactory,
+        this.validation = validation;
         InputFactory.Context<CollectExpression<T, ?>> ctx = inputFactory.ctxForRefs(txnCtx, refResolver);
         toInject = new HashMap<>();
+        nestedReferenceByTopLevelColumnName = new HashMap<>();
+
         for (Reference colWithDefault : allDefaultExpressionColumns) {
-            if (!presentColumns.contains(colWithDefault)) {
-                toInject.put(colWithDefault, ctx.add(colWithDefault.defaultExpression()));
-            }
+            toInject.put(colWithDefault, ctx.add(colWithDefault.defaultExpression()));
         }
+
         for (GeneratedReference generatedCol : allGeneratedColumns) {
-            if (presentColumns.contains(generatedCol)) {
-                if (validation == Validation.VALUE_MATCH) {
-                    toValidate.put(generatedCol, ctx.add(generatedCol.generatedExpression()));
-                }
-            } else {
-                toInject.put(generatedCol, ctx.add(generatedCol.generatedExpression()));
-            }
+            processGeneratedExpressionColumn(generatedCol, ctx.add(generatedCol.generatedExpression()));
         }
         expressions = ctx.expressions();
     }
 
+    private void processGeneratedExpressionColumn(Reference reference,
+                                                  Input<?> expressionInput) {
+        toInject.put(reference, expressionInput);
+        final ColumnIdent ident = reference.column();
+        if (!ident.isTopLevel()) {
+            List<Reference> references = nestedReferenceByTopLevelColumnName.getOrDefault(
+                ident.name(),
+                new ArrayList<>());
+            references.add(reference);
+            nestedReferenceByTopLevelColumnName.put(ident.name(), references);
+        }
+    }
+
     public void setNextRow(T row) {
         for (int i = 0; i < expressions.size(); i++) {
             expressions.get(i).setNextRow(row);
         }
     }
 
+    private void validateValue(Reference target, Object generatedValue, Object providedValue) {
+        //noinspection unchecked
+        if (target.valueType().compareValueTo(generatedValue, providedValue) != 0) {
+            throw new IllegalArgumentException(
+                "Given value " + providedValue +
+                " for generated column " + target.column() +
+                " does not match calculation " + ((GeneratedReference) target).formattedGeneratedExpression() + " = " +
+                generatedValue
+            );
+        }
+    }
+
     public void validateValue(Reference target, Object providedValue) {
-        Input<?> input = toValidate.get(target);
+        validateValue(target, providedValue, validation == Validation.VALUE_MATCH);
+    }
+
+    public void validateValue(Reference target, Object providedValue, boolean validate) {
+        if (providedValue == null) {
+            return;
+        }
+
+        Input<?> input = toInject.get(target);
```

Comment:
```suggestion
        Input<?> input = toInject.remove(target);
```

If I read the diff correctly currently `toInject.remove` is always called anyway if the key is present. So it is semantically the same.

Maybe add a short note why it is being removed?